### ISSUE TITLE
making archspec GPU-aware

### ISSUE THREAD

**Issue by boegel** [archspec/archspec#25]
We would like to make `archspec` aware of GPUs, probably starting with NVIDIA GPUs (since they're most prevalent in the HPC community currently).

Before diving in, I would like to collect some feedback/ideas in this issue.

For example, is running the `nvidia-smi` the best way to obtain information about the GPU(s) available?
How do we deal with having multiple different GPUs in a single system (do people do that?)?

Should we make `archspec` aware of which CUDA compute capabilities are supported by a GPU, and what the corresponding options for the `nvcc` compiler are?

**Comment by ajdecon** [comment]
Yes, running `nvidia-smi` is probably the easiest way to query information about the available GPUs. For feature detection, that tool doesn't report compute capability directly at this time, but you could easily build a lookup table similar to the [CPU table you pointed me to](https://github.com/archspec/archspec-json/blob/master/cpu/microarchitectures.json). There's a [table correlating compute capability to product name](https://developer.nvidia.com/cuda-gpus) on the NVIDIA developer site.

Alternatively, if you want to query the compute capability directly, you can do so using the CUDA API. There's a good example of how to do that in C++ in the [`deviceQuery` sample in the CUDA samples](https://github.com/NVIDIA/cuda-samples/blob/master/Samples/deviceQuery/deviceQuery.cpp).

You can also do this in Python with PyCUDA, for example:

```
$ cat get-compute-capability.py
import pycuda.autoinit
import pycuda.driver as cuda

for di in range(cuda.Device.count()):
    dev_attrs = cuda.Device(di).get_attributes()
    print("{}: {}.{}".format(
        di,
        dev_attrs[pycuda._driver.device_attribute.COMPUTE_CAPABILITY_MAJOR],
        dev_attrs[pycuda._driver.device_attribute.COMPUTE_CAPABILITY_MINOR]
    ))
$ nvidia-smi -L
GPU 0: Tesla V100-SXM2-32GB (UUID: GPU-...)
GPU 1: Tesla V100-SXM2-32GB (UUID: GPU-...)
$ python get-compute-capability.py
0: 7.0
1: 7.0
```

**Comment by boegel** [comment]
Thanks a lot for the feedback @ajdecon, really useful!

**Comment by cgmb** [comment]
LLVM has a pair of simple utilities to determine the GPU architecture: [amdgpu-arch](https://github.com/llvm/llvm-project/tree/009048810ac635a7ad6c5f788d537172418b6054/clang/tools/amdgpu-arch) [nvptx-arch](https://github.com/llvm/llvm-project/tree/009048810ac635a7ad6c5f788d537172418b6054/clang/tools/nvptx-arch)

nvptx-arch depends on libcuda. amdgpu-arch depends on libhsa-runtime64.
